Security/Core/Authentication/Provider/OAuthProvider.php

```diff
@@ -111,29 +109,51 @@ public function authenticate(TokenInterface $token)
         $this->userChecker->checkPreAuth($user);
         $this->userChecker->checkPostAuth($user);
 
-        $token = new OAuthToken($oldToken->getRawToken(), $user->getRoles());
-        $token->setResourceOwnerName($resourceOwner->getName());
-        $token->setUser($user);
-        $token->setAuthenticated(true);
-        $token->setRefreshToken($oldToken->getRefreshToken());
-        $token->setCreatedAt($oldToken->getCreatedAt());
-
-        return $token;
+        return $this->createOAuthToken($oldToken->getRawToken(), $oldToken, $user);
     }
 
     /**
-     * @return OAuthToken|TokenInterface
+     * @param OAuthToken $expiredToken
      */
-    protected function refreshToken(TokenInterface $expiredToken, ResourceOwnerInterface $resourceOwner)
+    private function refreshToken(TokenInterface $expiredToken, ResourceOwnerInterface $resourceOwner): OAuthToken
     {
         if (!$expiredToken->getRefreshToken()) {
             return $expiredToken;
         }
 
-        $token = new OAuthToken($resourceOwner->refreshAccessToken($expiredToken->getRefreshToken()));
-        $token->setRefreshToken($expiredToken->getRefreshToken());
+        /** @var UserInterface $user */
+        $user = $expiredToken->getUser();
+
+        $token = $this->createOAuthToken(
+            $resourceOwner->refreshAccessToken($expiredToken->getRefreshToken()),
+            $expiredToken,
+            $user
+        );
+
         $this->tokenStorage->setToken($token);
 
         return $token;
     }
+
+    /**
+     * @param string|array $data
+     */
+    private function createOAuthToken(
```

Comment:
Hi, may be protected here would be better? I have extended my Application to use 'RememberMeOAuthToken'. its the same like your OAuthToken (needs to be refreshed etc), but has an extra field.